### PR TITLE
Improve performance of emissary.output.DropOffUtil.processMetadata() by improving getCookedParameters.

### PR DESCRIPTION
**Pull request by jdcove2**
This method is called for each BaseDataObject and creates a new TreeMap each time along with a bunch of TreeMap$Entry objects. This can be reduced to reusing a single Map object with no Entry objects being created.

### CODE REVIEW

**Comment by jdcove2** [comment]
Calling BaseDataObject.getCookedParameters() creates a new TreeMap each time it is called which then creates an entry for each mapping. For the 20Meg mysqldump test file there are 4,433,080 maps and entries created for a total of 180,869,661 bytes. Reusing a single map that is backed by ArrayLists reduces this to "order 10".  This is shown in the following screen shots.

1. TreeMap objects created by old code.

image

2. TreeMap$Entry objects created by old code.

image

3. No TreeMap objects created by new code.

image

4. No TreeMap$Entry objects created by new code.

image

File: src/main/java/emissary/output/DropOffUtil.java
Comment (LightSeekerSC) [inline src/main/java/emissary/output/DropOffUtil.java:1002]: Why are we creating this map class? Also this doesn't seem to be a map. This has a lookup speed of O(n), whereas maps are expected to have a lookup of O(1)

File: src/main/java/emissary/output/DropOffUtil.java
Comment (LightSeekerSC) [inline src/main/java/emissary/output/DropOffUtil.java:1089]: Why not use a HashMap?

**Comment by jpdahlke** [comment]
Let's hold off on this one for now. Maintaining a custom data set for this application may be a bit much.

**Comment by LightSeekerSC** [comment]
> Let's hold off on this one for now. Maintaining a custom data set for this application may be a bit much.
> 
> Does it make sense that instead of building the whole TreeMap for filetype lookups, just have a getFileType method that performs the lookups of the keys necessary. It could look just like the current method but instead of a metaData.containsKey, do a p.getStringParameter().

I agree that this is the solution that we should pursue. Using getCookedParameters here is unnecessary (duplicating every single parameter to get a limited subset) and considering that this is critical path, we would do good in optimizing the getFileType() method to only check the desired fields.

**Review by fbruton — APPROVED** [review]

**Comment by jpdahlke** [comment]
Are we accepting the api breaking change for this?

**Comment by LightSeekerSC** [comment]
> Are we accepting the api breaking change for this?

Are you referring to the fact that the method signature is changing? I think that it is worthwhile, considering the substantial savings this can have at scale

Would this require us to do a major version update?

**Comment by drivenflywheel** [comment]
> > Are we accepting the api breaking change for this?
> 
> Are you referring to the fact that the method signature is changing? I think that it is worthwhile, considering the substantial savings this can have at scale
> 
> Would this require us to do a major version update?

While migrating to the new `DropOffUtil.getFileType` signatures, it might be wise to retain the existing ones and supporting methods, annotated as `@Deprecated`, because this API change will cause upstream build failures.  They should be easily resolved, but there won't be any choice about when to resolve them.  Deprecation would prevent the upstream build failures but still provide a path to leverage the changes via a follow-on ticket.

File: src/main/java/emissary/output/DropOffUtil.java
Comment (drivenflywheel) [inline src/main/java/emissary/output/DropOffUtil.java:718]: We should retain this old method signature and annotate it as `@Deprecated` to prevent upstream build failures.

File: src/main/java/emissary/output/DropOffUtil.java
Comment (drivenflywheel) [inline src/main/java/emissary/output/DropOffUtil.java:729]: Retain this old method signature and annotate it as `@Deprecated` to prevent upstream build failures.

**Review by drivenflywheel — CHANGES_REQUESTED** [review]
Need to resolve breaking API changes that will cause upstream build failures.

**Review by fbruton — CHANGES_REQUESTED** [review]
I agree that the original API call should be deprecated.

File: src/main/java/emissary/output/DropOffUtil.java
Comment (fbruton) [inline src/main/java/emissary/output/DropOffUtil.java:777]: Update method description by replacing metadata and form string with IBaseDataObject

File: src/main/java/emissary/output/DropOffUtil.java
Comment (fbruton) [inline src/main/java/emissary/output/DropOffUtil.java:787]: Update method description by replacing metadata with IBaseDataObject

File: src/test/java/emissary/output/DropOffUtilTest.java
Comment (fbruton) [inline src/test/java/emissary/output/DropOffUtilTest.java:467]: Recommend including the original test, since you're now deprecating the method instead of removing it.

**Review by fbruton — CHANGES_REQUESTED** [review]

**Review by fbruton — APPROVED** [review]

**Comment by jdcove2** [comment]
Observations:
1. getFileType(...,) should not modify BDO passed in.
2. getFileType(...) calculates fileType and completeFileType, so these should be split into two methods.

File: src/main/java/emissary/output/DropOffUtil.java
Comment (drivenflywheel) [inline src/main/java/emissary/output/DropOffUtil.java:841]: I'm a bit puzzled by this method name.  Why `getCompleteFileType` as the method name, but completely ignore any bdo parameter stored under the key `FileTypeCheckParameter.COMPLETE_FILETYPE.getFieldName()`?  Admittedly, these are a bit hard to grok since the documentation re: expected values stored under of various parameter keys is very sparse in this codebase.

**Review by fbruton — COMMENTED** [review]
I need help understanding the ```getCompleteFileType```? Does that need to be part of this merge request?

**Review by fbruton — CHANGES_REQUESTED** [review]
Recommend removing ```getCompleteFileType``` for now and including in a separate effort.

**Comment by jdcove2** [comment]
For systems that use Emissary and the current DropOffUtil.getFileType(...) method and care about whether the Map that is passed in contains the "COMPLETE_FILENAME" key/value pair after DropOffUtil.getFileType(...) is called will not function properly using only the new DropOffUtil.getFileType(...) method. The new DropOffUtil.getCompleteFileType(...) method is also needed to completely replace the current DropOffUtil.getFileType(...) method. Therefore the DropOffUtil.getCompleteFileType(...) method cannot be removed from this PR.

**Comment by drivenflywheel** [comment]
> For systems that use Emissary and the current DropOffUtil.getFileType(...) method and care about whether the Map that is passed in contains the "COMPLETE_FILENAME" key/value pair after DropOffUtil.getFileType(...) is called will not function properly using only the new DropOffUtil.getFileType(...) method. The new DropOffUtil.getCompleteFileType(...) method is also needed to completely replace the current DropOffUtil.getFileType(...) method. Therefore the DropOffUtil.getCompleteFileType(...) method cannot be removed from this PR.

Here's where I've been stuck:  I get that the new `getFileType()` overload no longer populates a value under the `COMPLETE_FILENAME.getFieldName()` key , and that leaves a gap for any downstream components looking for a value under that key.  

I don't see how the new `getCompleteFileType()` method addresses that. The method isn't populating that key, it just wraps the short algorithm to find the value of what would have been shoved into that key by the eliminated side-effect.  If  `getCompleteFileType()` exists so that a downstream component can call it and stuff its return value under the COMPLETE_FILETYPE key, then that component is where the getCompleteFileType() method should be defined and implemented. Its inclusion here is odd, and the downstream component(s) would already have to change.

**Comment by jdcove2** [comment]
After much discussion, it appears that the most support is behind leaving the new getFileType(...) and getCompleteFileType(...) the way they are, changing method names if desired. Then, add a utility method that completes the original getFileType(...) capability by doing the following:

1. takes an IBaseDataObject and a Map and returns void
2. calls the new getFileType(...) and getCompleteFileType(...) methods
3. applies the getFileType()/getCompleteFileType() results to the map, as appropriate.

**Review by fbruton — CHANGES_REQUESTED** [review]
My recommendation is to remove ```getCompleteFileType``` and add the following methods:
```public string getFileType(final IBaseDataObject bdo) // calls getAndPutFileType(bdo,null)```
```public string getAndPutFileType(final IBaseDataObject bdo, final Map<String, String> metaData) // derive fileType from bdo and put to metadata if metadata is not empty```
May need to include a method with ```formsArg```. I'm not sure if that is still necessary.

**Review by drivenflywheel — APPROVED** [review]
Approved, assuming the resolution of the most recent method names discussion.

**Review by fbruton — APPROVED** [review]

**Pull request merged**
